graviti/utility/user.py

```diff
@@ -355,3 +357,86 @@ def update(self, __m: Any = (), **kwargs: _V) -> None:  # pylint: disable=argume
 
         """
         self._data.update(__m, **kwargs)
+
+
+class StringKeyOrderedDict(MutableMapping[str, _V]):
+    """This class is a dict of ordered elements, supports searching the element by its index.
+
+    Arguments:
+        items: The items need to be stored into the StringKeyOrderedDict.
+
+    """
+
+    def __init__(
+        self, items: Union[Iterable[Tuple[str, _V]], Mapping[str, _V], None] = None
+    ) -> None:
+        self._keys: List[str] = []
+        self._data: Dict[str, _V] = {}
+
+        if not items:
+            return
+
+        iterable = items.items() if isinstance(items, Mapping) else items
+        for key, value in iterable:
+            self._keys.append(key)
+            self._data[key] = value
+
+    def __len__(self) -> int:
+        return self._data.__len__()
+
+    def __getitem__(self, key: Union[int, str]) -> _V:
+        if isinstance(key, int):
+            key = self._keys.__getitem__(key)
+
+        return self._data.__getitem__(key)
+
+    def __setitem__(self, key: Union[int, str], value: _V) -> None:
+        if isinstance(key, int):
+            key = self._keys.__getitem__(key)
+
+        if not self._data.__contains__(key):
+            self._keys.append(key)
+
+        self._data.__setitem__(key, value)
+
+    def __delitem__(self, key: Union[int, str]) -> None:
+        try:
+            if isinstance(key, int):
+                key = self._keys.pop(key)
+            else:
+                self._keys.remove(key)
+        except ValueError:
+            raise KeyError(key) from None
+
+        self._data.__delitem__(key)
+
+    def __contains__(self, key: Any) -> bool:
+        return self._data.__contains__(key)
+
+    def __iter__(self) -> Iterator[str]:
+        return self._keys.__iter__()
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, self.__class__):
+            return False
+
+        return self._keys.__eq__(other._keys) and self._data.__eq__(other._data)
+
+    def popitem(self) -> Tuple[str, _V]:
```

Comment:
why we need `popitem` here?
basicly `popitem` is a remove operation, but this `popitem` has not called `__delitem__`?

Reply:
Because if use the `MutableMapping.popitem` method, the first item will be popped. And we want it to popitem from the tail.

Reply:
Ok, we need to call `self.__delitem__` in popitem